File: 알고리즘스터디/jisu/programmers/Lv.1/크레인 인형뽑기 게임.js
const solution = (board, moves) => {
  let result = 0;
  let bucket = [];

  for (let m = 0; m < moves.length; m++) {
    let idx = moves[m] - 1;
    for (let i = 0; i < board.length; i++) {
      let boardRow = board[i];
      if (boardRow[idx] !== 0) {
        bucket.push(boardRow[idx]);
        boardRow[idx] = 0;
        break;
      }
    }
    let len = bucket.length;
    if (len > 1) {
      if (bucket[len - 1] === bucket[len - 2]) {
        bucket.splice(len - 2);
        result += 2;
      }
    }
  }

  return result;
};

File: 알고리즘스터디/hyein/programmers/크레인인형뽑기게임.java
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        Stack<Integer> basket = new Stack<>();
        int answer = 0;
        for(int move = 0; move < moves.length; move++) {
            int dollIdx = findDoll(board, moves[move]-1);
            if(dollIdx == -1) continue;
            
            int doll = board[dollIdx][moves[move]-1];
            board[dollIdx][moves[move]-1] = 0;
            
            if(!basket.isEmpty() && basket.peek() == doll) {
                answer += 2;
                basket.pop();
            }
            else basket.push(doll);
        }
        
        return answer;
    }
    
    private int findDoll(int[][] board, int move) {
        int idx = 0;
        while(idx < board.length && board[idx][move] == 0) idx++;
        
        if(idx == board.length) return -1;
        else return idx;
    }
}